src/Microsoft.SqlTools.ManagedBatchParser/ReliableConnection/SqlConnectionHelperScripts.cs

```diff
@@ -45,7 +45,21 @@ @rc 			int
 
 SELECT @filepath AS FilePath
 ";
-        public const string GetClusterEndpoints = @"SELECT [name], [description], [endpoint], [protocol_desc] FROM .[sys].[dm_cluster_endpoints];";
+        /**
+         * Query to get the server endpoints. We first try to query the dm_cluster_endpoints DMV since that will contain all
+         * of the endpoints. But if that fails (such as if the user doesn't have VIEW SERVER STATE permissions) then we'll
+         * fall back to just querying the ControllerEndpoint server property to at least get the endpoint of the controller
+         * and rely on the caller to connect to the controller to query for any of the other endpoints it needs.
+         */
+        public const string GetClusterEndpoints = @"BEGIN TRY
+SELECT [name], [description], [endpoint], [protocol_desc] FROM .[sys].[dm_cluster_endpoints]
+END TRY
+BEGIN CATCH
+DECLARE @endpoint VARCHAR(max)
+select @endpoint = CONVERT(VARCHAR(max),SERVERPROPERTY('ControllerEndpoint'))
```

Comment:
are other endpoints also exposed this way?

Reply:
No, just the controller. The idea is that you connect to the controller and then get the other endpoints that way.